File: src/analysis/spectral_mode.c
#include "algorithm.h"

#include "sounds/spectrum.h"

#include <stdlib.h>

typedef struct SpectrumModeAlgorithm {
    SoundSpectrumAnalyzer *spectrum;
    SoundSpectrumMode spectrum_mode;
    uint64_t column_samples;
    uint64_t last_center;
    uint64_t latency_samples;
} SpectrumModeAlgorithm;

static void spectrum_mode_destroy(void *state) {
    SpectrumModeAlgorithm *algorithm = state;

    if (!algorithm) {
        return;
    }

    sound_spectrum_analyzer_destroy(algorithm->spectrum);
    free(algorithm);
}

static double spectrum_mode_min_frequency(const void *state) {
    const SpectrumModeAlgorithm *algorithm = state;
    return sound_spectrum_analyzer_min_frequency(algorithm->spectrum);
}

static double spectrum_mode_max_frequency(const void *state) {
    const SpectrumModeAlgorithm *algorithm = state;
    return sound_spectrum_analyzer_max_frequency(algorithm->spectrum);
}

static void spectrum_mode_reset(void *state, uint64_t written_samples) {
    SpectrumModeAlgorithm *algorithm = state;

    algorithm->last_center = written_samples > algorithm->latency_samples ?
        written_samples - algorithm->latency_samples :
        0;
}

static bool spectrum_mode_update(
    void *state,
    const SoundAnalysisInput *input,
    SoundAnalysisOutput *output,
    bool emit,
    SoundError *error
) {
    SpectrumModeAlgorithm *algorithm = state;

    if (!emit) {
        return true;
    }

    uint64_t latest_center = input->written_samples > algorithm->latency_samples ?
        input->written_samples - algorithm->latency_samples :
        0;
    uint64_t first_center = algorithm->last_center + algorithm->column_samples;

    if (first_center < algorithm->latency_samples) {
        first_center = algorithm->latency_samples;
    }

    if (latest_center < first_center) {
        return true;
    }

    uint64_t pending = 1U + (latest_center - first_center) / algorithm->column_samples;
    if (pending > input->column_limit) {
        pending = input->column_limit;
    }

    for (uint64_t column = 0; column < pending; ++column) {
        uint64_t center_sample = first_center + column * algorithm->column_samples;
        float *dbfs_rows = NULL;

        if (!sound_analysis_output_append(output, &dbfs_rows, error)) {
            return false;
        }

        if (!sound_spectrum_analyzer_column_db(
                algorithm->spectrum,
                input->ring,
                center_sample,
                algorithm->spectrum_mode,
                dbfs_rows,
                input->row_count,
                error
            )) {
            return false;
        }

        algorithm->last_center = center_sample;
    }

    return true;
}

static const SoundAnalysisAlgorithmOps spectrum_mode_ops = {
    .destroy = spectrum_mode_destroy,
    .min_frequency = spectrum_mode_min_frequency,
    .max_frequency = spectrum_mode_max_frequency,
    .reset = spectrum_mode_reset,
    .toggle_sst = NULL,
    .sst_enabled = NULL,
    .update = spectrum_mode_update,
};

static bool spectrum_mode_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAppMode app_mode,
    SoundSpectrumMode spectrum_mode,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    *algorithm = (SoundAnalysisAlgorithm){0};

    SpectrumModeAlgorithm *state = calloc(1, sizeof(*state));
    if (!state) {
        sound_error_set(error, "could not allocate spectrum mode algorithm");
        return false;
    }

    state->column_samples = column_samples;
    state->spectrum_mode = spectrum_mode;

    if (!sound_spectrum_analyzer_create(
            sample_rate,
            columns_per_second,
            &state->spectrum,
            error
        )) {
        spectrum_mode_destroy(state);
        return false;
    }

    state->latency_samples = sound_spectrum_analyzer_latency_samples(state->spectrum);

    *algorithm = (SoundAnalysisAlgorithm){
        .ops = &spectrum_mode_ops,
        .state = state,
        .mode = app_mode,
    };
    return true;
}

bool sound_reassigned_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_REASSIGNED,
        SOUND_SPECTRUM_REASSIGNED,
        algorithm,
        error
    );
}

bool sound_squeezed_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_SQUEEZED,
        SOUND_SPECTRUM_SQUEEZED,
        algorithm,
        error
    );
}

bool sound_superlet_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_SUPERLET,
        SOUND_SPECTRUM_SUPERLET,
        algorithm,
        error
    );
}

bool sound_multitaper_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_MULTITAPER,
        SOUND_SPECTRUM_MULTITAPER,
        algorithm,
        error
    );
}

bool sound_s_transform_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_S_TRANSFORM,
        SOUND_SPECTRUM_S_TRANSFORM,
        algorithm,
        error
    );
}

bool sound_sparse_algorithm_create(
    double sample_rate,
    double columns_per_second,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    return spectrum_mode_algorithm_create(
        sample_rate,
        columns_per_second,
        column_samples,
        SOUND_APP_MODE_SPARSE,
        SOUND_SPECTRUM_SPARSE,
        algorithm,
        error
    );
}

File: src/analysis/spectral_mode.zig
//! Sparse ridges, reassigned, squeezed, superlet, multitaper, S-transform modes.
//! Rewrite target; C reference: src_c/src/analysis/spectral_mode.c

const std = @import("std");
const spectrum = @import("spectrum.zig");
const ring_buffer = @import("../audio/ring_buffer.zig");

pub const Error = error{
    OutputFull,
};

pub const Algorithm = struct {
    spectrum_analyzer: spectrum.Analyzer,
    spectrum_mode: spectrum.Mode,
    column_samples: u64,
    last_center: u64,
    latency_samples: u64,

    pub fn init(
        allocator: std.mem.Allocator,
        sample_rate: f64,
        columns_per_second: f64,
        column_samples: u64,
        mode: spectrum.Mode,
    ) !Algorithm {
        var analyzer = try spectrum.Analyzer.init(allocator, sample_rate, columns_per_second);
        errdefer analyzer.deinit();

        return .{
            .spectrum_analyzer = analyzer,
            .spectrum_mode = mode,
            .column_samples = column_samples,
            .last_center = 0,
            .latency_samples = analyzer.latencySamples(),
        };
    }

    pub fn deinit(self: *Algorithm) void {
        self.spectrum_analyzer.deinit();
        self.* = undefined;
    }

    pub fn minFrequency(self: *const Algorithm) f64 {
        return self.spectrum_analyzer.minFrequency();
    }

    pub fn maxFrequency(self: *const Algorithm) f64 {
        return self.spectrum_analyzer.maxFrequency();
    }

    pub fn setFrequencyRange(self: *Algorithm, min_hz: f64, max_hz: f64) !void {
        try self.spectrum_analyzer.setFrequencyRange(min_hz, max_hz);
    }

    pub fn reset(self: *Algorithm, written_samples: u64) void {
        self.last_center = if (written_samples > self.latency_samples)
            written_samples - self.latency_samples
        else
            0;
    }

    pub fn update(
        self: *Algorithm,
        ring: *const ring_buffer.RingBuffer,
        written_samples: u64,
        row_count: usize,
        column_limit: usize,
        columns: []f32,
        column_count: *usize,
    ) !void {
        if (row_count == 0 or column_limit == 0) return;

        const latest_center = if (written_samples > self.latency_samples)
            written_samples - self.latency_samples
        else
            0;
        var first_center = self.last_center + self.column_samples;

        if (first_center < self.latency_samples) first_center = self.latency_samples;
        if (latest_center < first_center) return;

        var pending = 1 + (latest_center - first_center) / self.column_samples;
        pending = @min(pending, @as(u64, @intCast(column_limit)));

        for (0..@intCast(pending)) |column| {
            const center_sample = first_center + @as(u64, @intCast(column)) * self.column_samples;
            const dbfs_rows = try appendColumn(columns, row_count, column_limit, column_count);

            try self.spectrum_analyzer.columnDb(
                ring,
                center_sample,
                self.spectrum_mode,
                dbfs_rows,
            );
            self.last_center = center_sample;
        }
    }
};

fn appendColumn(columns: []f32, row_count: usize, capacity: usize, column_count: *usize) ![]f32 {
    if (column_count.* >= capacity) return Error.OutputFull;

    const start = column_count.* * row_count;
    column_count.* += 1;
    return columns[start..][0..row_count];
}
